pkg/ebpf/tcp_queue_length.go

```diff
@@ -0,0 +1,154 @@
+		return nil, fmt.Errorf("Couldn’t find asset “tcp-queue-length-kern.c”: %v", err)
+	}
+
+	// Process the `#include` of embedded headers.
+	// Note that embedded headers including other embedded headers is not managed because
+	// this would also require to properly handle inclusion guards.
+	includeRegexp := regexp.MustCompile(`^\s*#\s*include\s+"(.*)"$`)
+	var source bytes.Buffer
+	scanner := bufio.NewScanner(bytes.NewBuffer(source_raw))
+	for scanner.Scan() {
+		match := includeRegexp.FindSubmatch(scanner.Bytes())
+		if len(match) == 2 {
+			header, err := Asset(string(match[1]))
+			if err == nil {
+				source.Write(header)
+				continue
+			}
+		}
+		source.Write(scanner.Bytes())
+		source.WriteByte('\n')
+	}
+
+	m := bpflib.NewModule(source.String(), []string{})
+	if m == nil {
+		return nil, fmt.Errorf("Failed to compile “tcp-queue-length-kern.c”")
+	}
+
+	kprobe_recvmsg, err := m.LoadKprobe("kprobe__tcp_recvmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKprobe("tcp_recvmsg", kprobe_recvmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_recvmsg: %s\n", err)
+	}
+
+	kretprobe_recvmsg, err := m.LoadKprobe("kretprobe__tcp_recvmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kretprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKretprobe("tcp_recvmsg", kretprobe_recvmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_recvmsg: %s\n", err)
+	}
+
+	kprobe_sendmsg, err := m.LoadKprobe("kprobe__tcp_sendmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKprobe("tcp_sendmsg", kprobe_sendmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_sendmsg: %s\n", err)
+	}
+
+	kretprobe_sendmsg, err := m.LoadKprobe("kretprobe__tcp_sendmsg")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load kretprobe__tcp_recvmsg: %s\n", err)
+	}
+
+	if err := m.AttachKretprobe("tcp_sendmsg", kretprobe_sendmsg, -1); err != nil {
+		return nil, fmt.Errorf("Failed to attach tcp_sendmsg: %s\n", err)
+	}
+
+	table := bpflib.NewTable(m.TableId("queue"), m)
+
+	return &TCPQueueLengthTracer{
+		m:        m,
+		queueMap: table,
+	}, nil
+}
+
+func (t *TCPQueueLengthTracer) Close() {
+	t.m.Close()
+}
+
+func (t *TCPQueueLengthTracer) Get() []tcpqueuelength.Stats {
+	if t == nil {
+		return nil
+	}
+
+	var result []tcpqueuelength.Stats
+
+	for it := t.queueMap.Iter(); it.Next(); {
+		var in C.struct_stats        // kernel       <-> system-probe
+		var out tcpqueuelength.Stats // system-probe <-> agent
+
+		// `binary.Read(…)` doesn’t work because reflection doesn’t work with C types.
+		// binary.Read(bytes.NewBuffer(it.Leaf()), bpflib.GetHostByteOrder(), &in)
+
+		data := it.Leaf()
+		C.memcpy(unsafe.Pointer(&in), unsafe.Pointer(&data[0]), C.sizeof_struct_stats)
+
+		// TODO: Can this code be handled by using reflection? Would it be clearer?
```

Comment:
Could we factor the conversion from `C.struct_stats` to ` tcpqueuelength.Stats` into a dedicated function instead of doing it inside `Get()`.

Reply:
Good suggestion !
Done !